feature/experimental/p4rt/ate_tests/traceroute_packetin_test/packetin_test.go

```diff
@@ -0,0 +1,255 @@
+
+	ate.Traffic().Stop(t)
+}
+
+// fetchPackets reads p4rt packets sent to p4rt client.
+func fetchPackets(ctx context.Context, t *testing.T, client *p4rt_client.P4RTClient, expectNumber int) []*p4rt_client.P4RTPacketInfo {
+	t.Helper()
+	packets := []*p4rt_client.P4RTPacketInfo{}
+	for i := 0; i < expectNumber; i++ {
+		_, packet, err := client.StreamChannelGetPacket(&streamName, 0)
+		if err == io.EOF {
+			t.Logf("EOF error is seen in PacketIn.")
+			break
+		} else if err == nil {
+			if packet != nil {
+				packets = append(packets, packet)
+			}
+		} else {
+			t.Fatalf("There is error seen when receving packets. %v, %s", err, err)
+			break
+		}
+	}
+	return packets
+}
+
+// testPacketIn programs p4rt table entry and sends traffic related to Traceroute,
+// then validates packetin message metadata and payload.
+func testPacketIn(ctx context.Context, t *testing.T, args *testArgs, IsIpv4 bool) {
+	leader := args.leader
+	follower := args.follower
+
+	if IsIpv4 {
+		// Insert p4rtutils acl entry on the DUT
+		if err := programmTableEntry(ctx, t, leader, args.packetIO, false, true); err != nil {
+			t.Fatalf("There is error when programming entry")
+		}
+		// Delete p4rtutils acl entry on the device
+		defer programmTableEntry(ctx, t, leader, args.packetIO, true, true)
+	} else {
+		// Insert p4rtutils acl entry on the DUT
+		if err := programmTableEntry(ctx, t, leader, args.packetIO, false, false); err != nil {
+			t.Fatalf("There is error when programming entry")
+		}
+		// Delete p4rtutils acl entry on the device
+		defer programmTableEntry(ctx, t, leader, args.packetIO, true, false)
+	}
+
+	// Send GDP traffic from ATE
+	srcEndPoint := args.top.Interfaces()[atePort1.Name]
+
+	packetInTests := []struct {
+		desc       string
+		client     *p4rt_client.P4RTClient
+		expectPass bool
+	}{{
+		desc:       "PacketIn to Primary Controller",
+		client:     leader,
+		expectPass: true,
+	}, {
+		desc:       "PacketIn to Secondary Controller",
+		client:     follower,
+		expectPass: false,
+	}}
+
+	//CheckTTL struct for TTL/HopLimit=0,1
+	checkTTL := []struct {
+		desc string
+		TTL  uint8
+	}{{
+		desc: "TTL/HopLimit 1",
+		TTL:  1,
+	}, {
+		desc: "TTL/HopLimit 0",
+		TTL:  0,
+	}}
+
+	for _, TTL := range checkTTL {
+		t.Log(TTL.desc)
+		testTraffic(t, args.ate, args.packetIO.GetTrafficFlow(args.ate, IsIpv4, TTL.TTL, 300, 2), srcEndPoint, 10)
+		for _, test := range packetInTests {
+			t.Run(test.desc, func(t *testing.T) {
+				// Extract packets from PacketIn message sent to p4rt client
+				packets := fetchPackets(ctx, t, test.client, 40)
+
+				if !test.expectPass {
+					if len(packets) > 0 {
+						t.Fatalf("Unexpected packets received.")
+					}
+				} else {
+					if len(packets) == 0 {
+						t.Fatalf("There are no packets received.")
+					}
+					t.Logf("Start to decode packet and compare with expected packets.")
+					wantPacket := args.packetIO.GetPacketTemplate()
+					for _, packet := range packets {
+						if packet != nil {
```

Comment:
if packet == nil  { continue }

Reply:
This condition is already taken care at the beginning of 'packetInTests' loop. 
if len(packets) == 0 {}
Test would Fatal fail if this triggers.